src/Symfony/Component/Mailer/Bridge/MicrosoftGraph/Transport/MicrosoftGraphTransportFactory.php

```diff
@@ -0,0 +1,65 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Mailer\Bridge\MicrosoftGraph\Transport;
+
+use Symfony\Component\Mailer\Bridge\MicrosoftGraph\TokenManager;
+use Symfony\Component\Mailer\Exception\IncompleteDsnException;
+use Symfony\Component\Mailer\Exception\InvalidArgumentException;
+use Symfony\Component\Mailer\Exception\UnsupportedSchemeException;
+use Symfony\Component\Mailer\Transport\AbstractTransportFactory;
+use Symfony\Component\Mailer\Transport\Dsn;
+use Symfony\Component\Mailer\Transport\TransportInterface;
+
+class MicrosoftGraphTransportFactory extends AbstractTransportFactory
+{
+    public function create(Dsn $dsn): TransportInterface
+    {
+        if ('microsoft+graphapi' !== $dsn->getScheme()) {
+            throw new UnsupportedSchemeException($dsn, 'microsoft graph api', $this->getSupportedSchemes());
+        }
+
+        $tenantId = $dsn->getOption('tenantId');
+        if (null === $tenantId) {
+            throw new IncompleteDsnException("Transport 'microsoft+graphapi' requires the 'tenant' option.");
+        }
+
+        $graphEndpoint = $dsn->getHost();
+        $authEndpoint = $dsn->getOption('authEndpoint');
+        if ('default' === $graphEndpoint) {
+            $graphEndpoint = 'graph.microsoft.com';
+            if (null === $authEndpoint) {
+                $authEndpoint = 'login.microsoftonline.com';
+            }
+        }
+
+        if (null === $authEndpoint) {
+            throw new IncompleteDsnException("Transport 'microsoft+graphapi' requires the 'authEndpoint' option when not using the default graph endpoint.");
```

Comment:
why? can't we default to login.microsoftonline.com (that's what my previous suggestion proposes, but maybe it doesn't make sense)

Reply:
If you are explicitly not using the default graph endpoint (MS global), you will need to use a different auth endpoint as well (unless you set the graph endpoint to its default value without using the `default` keyword). See https://learn.microsoft.com/en-us/graph/deployments#app-registration-and-token-service-root-endpoints.

So, for me this is more of a "you probably made a configuration error" error message, as you won't be able to get a token when you're using the wrong auth endpoint.

Reply:
I'm fine as is, thanks for the details